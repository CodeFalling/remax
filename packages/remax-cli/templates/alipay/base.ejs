<template name="REMAX_TPL">
  <block a:for="{{root.children}}" a:key="{{item.id}}">
    <template is="REMAX_TPL_1" data="{{item}}"/>
  </block>
</template>

<template name="REMAX_TPL_1">
  <block>
    <block a:if="{{item && item.text}}">
      <!-- string -->
      {{item.text}}
    </block>

    <% for (let component of components) { %>
      <%- include('./component.ejs', {
        type: component.type,
        props: component.props,
        id: component.id,
      }) %>
    <% } %>
  </block>
</template>
