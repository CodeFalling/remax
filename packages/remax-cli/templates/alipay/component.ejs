<% camelCased = function(str) {
  if (str.indexOf('-') !== -1) {
    return str;
  }
  return str.replace(/-([a-z0-9])/g, function (g) { return g[1].toUpperCase();
})}%>

<template name="REMAX_TPL_1_<%=id%>">
  <<%=id%> <% for(let key of props) { %> <%=key%>="{{item.props['<%=camelCased(key)%>']}}" <% } %>>
    <% if (id === 'swiper') { %>
    <block a:for="{{item.children}}" a:key="{{item.id}}">
      <swiper-item key="{{item.props.key}}">
        <block a:for="{{item.children}}" a:key="{{item.id}}">
          <template is="REMAX_TPL_1_CONTAINER" data="{{item: item}}" />
        </block>
      </swiper-item>
    </block>
    <% } else if (id === 'picker') { %>
    <view>
      <template is="REMAX_TPL_1_CONTAINER" data="{{item: item.children[0]}}" />
    </view>
    <% } else if (id === 'picker-view') { %>
    <block a:for="{{item.children}}" a:key="{{item.id}}">
      <picker-view-column>
        <view a:for="{{item.children}}">
          <template is="REMAX_TPL_1_CONTAINER" data="{{item}}" />
        </view>
      </picker-view-column>
    </block>
    <% } else { %>
    <block a:for="{{item.children}}" a:key="{{item.id}}">
      <template is="REMAX_TPL_1_CONTAINER" data="{{item: item}}" />
    </block>
    <% } %>
  </<%=id%>>
</template>