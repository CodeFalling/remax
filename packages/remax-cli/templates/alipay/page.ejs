<import-sjs name="helper" from="<%= jsHelper %>"/>
<import src="<%= baseTemplate %>"/>
<template is="REMAX_TPL" data="{{tree: helper.reduce(action)}}"/>
