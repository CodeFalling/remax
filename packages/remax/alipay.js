export * from './esm/adapters/alipay';
