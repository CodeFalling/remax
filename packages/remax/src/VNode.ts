import { TYPE_TEXT } from './constants';
import Container from './Container';

export interface RawNode {
  id?: number;
  type: string | Symbol;
  props?: any;
  children?: RawNode[];
  text?: string;
}

export type Path = Array<string | number>;

export default class VNode {
  id: number;
  container: Container;
  children?: VNode[];
  mounted = false;
  type: string | Symbol;
  props: any;
  parent: VNode | null = null;
  text?: string;

  constructor(id: number, type: string | Symbol, props: any, page: any) {
    this.id = id;
    this.container = page;
    this.type = type;
    this.props = props;
  }

  appendChild(node: VNode) {
    node.parent = this;
    this.children = this.children || [];
    this.children.push(node);
    if (this.isMounted()) {
      this.container.requestUpdate(node.path(), node.toJSON());
    }
  }

  removeChild(node: VNode) {
    const start = this.children!.indexOf(node);
    this.children!.splice(start, 1);
    if (this.isMounted()) {
      const path = this.parent
        ? this.parent.path()
        : [...this.path(), 'children'];
      this.container.requestSpliceUpdate(path, start, 1);
    }
  }

  insertBefore(newNode: VNode, referenceNode: VNode) {
    newNode.parent = this;
    const start = this.children!.indexOf(referenceNode);
    this.children!.splice(start, 0, newNode);
    if (this.isMounted()) {
      const path = this.parent
        ? this.parent.path()
        : [...this.path(), 'children'];
      this.container.requestSpliceUpdate(path, start, 0, newNode);
    }
  }

  path(): Path {
    if (!this.parent) {
      return [this.container.root.length - 1];
    }
    return [
      ...this.parent.path(),
      'children',
      this.parent.children!.indexOf(this),
    ];
  }

  isMounted(): boolean {
    return this.parent ? this.parent.isMounted() : this.mounted;
  }

  toJSON(): RawNode {
    if (this.type === TYPE_TEXT) {
      return {
        type: this.type,
        text: this.text,
      };
    }
    return {
      id: this.id,
      type: this.type,
      props: this.props,
      children: this.children && this.children.map(c => c.toJSON()),
      text: this.text,
    };
  }
}
