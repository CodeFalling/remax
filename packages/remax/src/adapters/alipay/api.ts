// TODO promise 化补充完整
import promisify from '../../utils/promisify';

export const getAppStub = getApp;

export const addCardAuth = my.addCardAuth;
export const addPhoneContact = my.addPhoneContact;
export const alert = my.alert;
export const canIUse = my.canIUse;
export const chooseAlipayContact = my.chooseAlipayContact;
export const chooseCity = my.chooseCity;
export const chooseContact = my.chooseContact;
export const chooseImage = promisify(my.chooseImage);
export const chooseLocation = my.chooseLocation;
export const choosePhoneContact = my.choosePhoneContact;
export const clearStorage = my.clearStorage;
export const clearStorageSync = my.clearStorageSync;
export const closeBluetoothAdapter = my.closeBluetoothAdapter;
export const closeSocket = my.closeSocket;
export const compressImage = my.compressImage;
export const confirm = my.confirm;
export const connectBLEDevice = my.connectBLEDevice;
export const connectSocket = my.connectSocket;
export const createAnimation = my.createAnimation;
export const createCanvasContext = my.createCanvasContext;
export const createIntersectionObserver = (my as any)
  .createIntersectionObserver;
export const createMapContext = my.createMapContext;
export const createSelectorQuery = my.createSelectorQuery;
export const createWebViewContext = my.createWebViewContext;
export const datePicker = my.datePicker;
export const disconnectBLEDevice = my.disconnectBLEDevice;
export const downloadFile = my.downloadFile;
export const getAuthCode = my.getAuthCode;
export const getAuthUserInfo = my.getAuthUserInfo;
export const getBatteryInfo = (my as any).getBatteryInfo;
export const getBatteryInfoSync = (my as any).getBatteryInfoSync;
export const getBeacons = my.getBeacons;
export const getBLEDeviceCharacteristics = my.getBLEDeviceCharacteristics;
export const getBLEDeviceServices = my.getBLEDeviceServices;
export const getBluetoothAdapterState = my.getBluetoothAdapterState;
export const getBluetoothDevices = my.getBluetoothDevices;
export const getClipboard = my.getClipboard;
export const getConnectedBluetoothDevices = my.getConnectedBluetoothDevices;
export const getFileInfo = my.getFileInfo;
export const getImageInfo = my.getImageInfo;
export const getLocation = my.getLocation;
export const getNetworkType = my.getNetworkType;
export const getPhoneNumber = my.getPhoneNumber;
export const getRunData = (my as any).getRunData;
export const getRunScene = my.getRunScene;
export const getSavedFileInfo = my.getSavedFileInfo;
export const getSavedFileList = my.getSavedFileList;
export const getScreenBrightness = my.getScreenBrightness;
export const getServerTime = my.getServerTime;
export const getSetting = my.getSetting;
export const getStorage = my.getStorage;
export const getStorageInfo = my.getStorageInfo;
export const getStorageInfoSync = my.getStorageInfoSync;
export const getStorageSync = my.getStorageSync;
export const getSystemInfo = my.getSystemInfo;
export const getSystemInfoSync = my.getSystemInfoSync;
export const getTitleColor = (my as any).getTitleColor;
export const getUpdateManager = (my as any).getUpdateManager;
export const hideAddToDesktopMenu = my.hideAddToDesktopMenu;
export const hideAllAddToDesktopMenu = my.hideAllAddToDesktopMenu;
export const hideAllFavoriteMenu = my.hideAllFavoriteMenu;
export const hideBackHome = my.hideBackHome;
export const hideFavoriteMenu = my.hideFavoriteMenu;
export const hideKeyboard = my.hideKeyboard;
export const hideLoading = my.hideLoading;
export const hideNavigationBarLoading = my.hideNavigationBarLoading;
export const hideShareMenu = my.hideShareMenu;
export const hideTabBar = my.hideTabBar;
export const hideTabBarRedDot = my.hideTabBarRedDot;
export const hideToast = my.hideToast;
export const loadFontFace = (my as any).loadFontFace;
export const makePhoneCall = my.makePhoneCall;
export const multiLevelSelect = my.multiLevelSelect;
export const navigateBack = promisify(my.navigateBack);
export const navigateBackMiniProgram = promisify(my.navigateBackMiniProgram);
export const navigateTo = promisify(my.navigateTo);
export const navigateToMiniProgram = my.navigateToMiniProgram;
export const notifyBLECharacteristicValueChange =
  my.notifyBLECharacteristicValueChange;
export const offAccelerometerChange = my.offAccelerometerChange;
export const offBLECharacteristicValueChange =
  my.offBLECharacteristicValueChange;
export const offBLEConnectionStateChanged = my.offBLEConnectionStateChanged;
export const offBluetoothAdapterStateChange = my.offBluetoothAdapterStateChange;
export const offBluetoothDeviceFound = my.offBluetoothDeviceFound;
export const offCompassChange = my.offCompassChange;
export const offGyroscopeChange = my.offGyroscopeChange;
export const offMemoryWarning = (my as any).offMemoryWarning;
export const offNetworkStatusChange = my.offNetworkStatusChange;
export const offSocketClose = my.offSocketClose;
export const offSocketError = (my as any).offSocketError;
export const offSocketMessage = (my as any).offSocketMessage;
export const offSocketOpen = (my as any).offSocketOpen;
export const offUserCaptureScreen = my.offUserCaptureScreen;
export const onAccelerometerChange = my.onAccelerometerChange;
export const onBeaconServiceChange = my.onBeaconServiceChange;
export const onBeaconUpdate = my.onBeaconUpdate;
export const onBLECharacteristicValueChange = my.onBLECharacteristicValueChange;
export const onBLEConnectionStateChanged = my.onBLEConnectionStateChanged;
export const onBluetoothAdapterStateChange = my.onBluetoothAdapterStateChange;
export const onBluetoothDeviceFound = my.onBluetoothDeviceFound;
export const onCompassChange = my.onCompassChange;
export const onGyroscopeChange = my.onGyroscopeChange;
export const onMemoryWarning = (my as any).onMemoryWarning;
export const onNetworkStatusChange = my.onNetworkStatusChange;
export const onSocketClose = my.onSocketClose;
export const onSocketError = my.onSocketError;
export const onSocketMessage = my.onSocketMessage;
export const onSocketOpen = my.onSocketOpen;
export const onUserCaptureScreen = my.onUserCaptureScreen;
export const openBluetoothAdapter = my.openBluetoothAdapter;
export const openCardDetail = my.openCardDetail;
export const openCardList = my.openCardList;
export const openKBVoucherDetail = my.openKBVoucherDetail;
export const openLocation = my.openLocation;
export const openMerchantCardList = my.openMerchantCardList;
export const openMerchantTicketList = my.openMerchantTicketList;
export const openMerchantVoucherList = my.openMerchantVoucherList;
export const openSetting = my.openSetting;
export const openTicketDetail = my.openTicketDetail;
export const openTicketList = my.openTicketList;
export const openVoucherDetail = my.openVoucherDetail;
export const openVoucherList = my.openVoucherList;
export const optionsSelect = my.optionsSelect;
export const pageScrollTo = my.pageScrollTo;
export const previewImage = my.previewImage;
export const prompt = my.prompt;
export const readBLECharacteristicValue = my.readBLECharacteristicValue;
export const redirectTo = my.redirectTo;
export const reLaunch = my.reLaunch;
export const removeSavedFile = my.removeSavedFile;
export const removeStorage = my.removeStorage;
export const removeStorageSync = my.removeStorageSync;
export const removeTabBarBadge = my.removeTabBarBadge;
export const reportAnalytics = my.reportAnalytics;
export const request = my.request;
export const rsa = my.rsa;
export const saveFile = my.saveFile;
export const saveImage = my.saveImage;
export const scan = my.scan;
export const SDKVersion = my.SDKVersion;
export const sendSocketMessage = my.sendSocketMessage;
export const setBackgroundColor = (my as any).setBackgroundColor;
export const setBackgroundTextStyle = (my as any).setBackgroundTextStyle;
export const setCanPullDown = (my as any).setCanPullDown;
export const setClipboard = my.setClipboard;
export const setKeepScreenOn = my.setKeepScreenOn;
export const setNavigationBar = my.setNavigationBar;
export const setOptionMenu = (my as any).setOptionMenu;
export const setScreenBrightness = my.setScreenBrightness;
export const setStorage = my.setStorage;
export const setStorageSync = my.setStorageSync;
export const setTabBarBadge = my.setTabBarBadge;
export const setTabBarItem = my.setTabBarItem;
export const setTabBarStyle = my.setTabBarStyle;
export const showActionSheet = my.showActionSheet;
export const showAuthGuide = my.showAuthGuide;
export const showLoading = my.showLoading;
export const showNavigationBarLoading = my.showNavigationBarLoading;
export const showSharePanel = (my as any).showSharePanel;
export const showTabBar = my.showTabBar;
export const showTabBarRedDot = my.showTabBarRedDot;
export const showToast = my.showToast;
export const startBeaconDiscovery = my.startBeaconDiscovery;
export const startBluetoothDevicesDiscovery = my.startBluetoothDevicesDiscovery;
export const startPullDownRefresh = my.startPullDownRefresh;
export const startZMVerify = my.startZMVerify;
export const stopBeaconDiscovery = my.stopBeaconDiscovery;
export const stopBluetoothDevicesDiscovery = my.stopBluetoothDevicesDiscovery;
export const stopPullDownRefresh = my.stopPullDownRefresh;
export const switchTab = my.switchTab;
export const textRiskIdentification = my.textRiskIdentification;
export const tradePay = my.tradePay;
export const uploadFile = my.uploadFile;
export const vibrate = my.vibrate;
export const vibrateLong = my.vibrateLong;
export const vibrateShort = my.vibrateShort;
export const watchShake = my.watchShake;
export const writeBLECharacteristicValue = my.writeBLECharacteristicValue;
