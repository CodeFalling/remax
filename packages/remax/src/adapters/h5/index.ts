export { default as createAppConfig } from './createAppConfig';
export { default as createPageConfig } from './createPageConfig';
export { default as Platform } from '../../Platform';
export * from './components';
export * from './api';
