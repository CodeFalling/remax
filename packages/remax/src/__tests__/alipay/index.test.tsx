import * as React from 'react';
import './helpers/setupGlobals';
import { render, View } from '../../../src/adapters/alipay';
import { reset } from '../../instanceId';
import VNode from '../../VNode';

class Context {
  root: VNode[] = [];

  updateData() {}

  appendChild(child: VNode) {
    this.root.push(child);
  }

  removeChild(child: VNode) {
    const start = this.root.indexOf(child);
    if (start >= 0) {
      this.root.splice(start, 1);
    }
  }

  insertBefore(child: VNode, beforeChild: VNode) {
    const start = this.root.indexOf(beforeChild);
    this.root.splice(start, 0, child);
  }
}

describe('remax render', () => {
  afterEach(() => {
    reset();
  });

  it('render correctly', () => {
    const Page = () => <View className="foo">hello</View>;
    const context = new Context();
    render(<Page />, context);
    expect(context.root).toMatchSnapshot();
  });

  it('insert new element', () => {
    class Page extends React.Component {
      state = {
        list: [1, 3],
      };

      update() {
        this.setState({
          list: [1, 2, 3],
        });
      }

      render() {
        const { list } = this.state;
        return (
          <View>
            {list.map(i => (
              <View key={i}>{i}</View>
            ))}
          </View>
        );
      }
    }

    const context = new Context();
    const page = React.createRef<any>();
    render(<Page ref={page} />, context);
    expect(context.root).toMatchSnapshot();
    page.current.update();
    expect(context.root).toMatchSnapshot();
  });

  it('umount component', () => {
    class Page extends React.Component {
      state = {
        show: true,
      };

      hide() {
        this.setState({ show: false });
      }

      render() {
        return <View>{this.state.show && <View>foo</View>}</View>;
      }
    }
    const context = new Context();
    const page = React.createRef<any>();
    render(<Page ref={page} />, context);
    expect(context.root).toMatchSnapshot();
    page.current.hide();
    expect(context.root).toMatchSnapshot();
  });

  it('renders style', () => {
    const Page = () => (
      <View style={{ width: '100px', height: '100px' }}>hello</View>
    );
    const context = new Context();
    render(<Page />, context);
    expect(context.root).toMatchSnapshot();
    const style = context.root[0].children![0].props.style;
    expect(style[style.length - 1]).toBe(';');
  });

  it('renders conditional fragment correctly', () => {
    class Page extends React.Component {
      state = {
        show: false,
      };

      show() {
        this.setState({ show: true });
      }

      render() {
        if (this.state.show) {
          return <View>foo</View>;
        }

        return (
          <>
            <View>one</View>
            <View>two</View>
          </>
        );
      }
    }
    const context = new Context();
    const page = React.createRef<any>();
    render(<Page ref={page} />, context);
    expect(context.root).toMatchSnapshot();
    page.current.show();
    expect(context.root).toMatchSnapshot();
  });
});
