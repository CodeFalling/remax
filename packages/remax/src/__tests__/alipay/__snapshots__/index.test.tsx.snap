// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`remax render insert new element 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "text": "1",
            "type": Symbol(text),
          },
        ],
        "id": 1,
        "props": Object {},
        "text": undefined,
        "type": "view",
      },
      Object {
        "children": Array [
          Object {
            "text": "3",
            "type": Symbol(text),
          },
        ],
        "id": 3,
        "props": Object {},
        "text": undefined,
        "type": "view",
      },
    ],
    "id": 4,
    "props": Object {},
    "text": undefined,
    "type": "view",
  },
]
`;

exports[`remax render insert new element 2`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "text": "1",
            "type": Symbol(text),
          },
        ],
        "id": 1,
        "props": Object {},
        "text": undefined,
        "type": "view",
      },
      Object {
        "children": Array [
          Object {
            "text": "2",
            "type": Symbol(text),
          },
        ],
        "id": 6,
        "props": Object {},
        "text": undefined,
        "type": "view",
      },
      Object {
        "children": Array [
          Object {
            "text": "3",
            "type": Symbol(text),
          },
        ],
        "id": 3,
        "props": Object {},
        "text": undefined,
        "type": "view",
      },
    ],
    "id": 4,
    "props": Object {},
    "text": undefined,
    "type": "view",
  },
]
`;

exports[`remax render render correctly 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "hello",
        "type": Symbol(text),
      },
    ],
    "id": 1,
    "props": Object {
      "class": "foo",
    },
    "text": undefined,
    "type": "view",
  },
]
`;

exports[`remax render renders conditional fragment correctly 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "one",
        "type": Symbol(text),
      },
    ],
    "id": 1,
    "props": Object {},
    "text": undefined,
    "type": "view",
  },
  Object {
    "children": Array [
      Object {
        "text": "two",
        "type": Symbol(text),
      },
    ],
    "id": 3,
    "props": Object {},
    "text": undefined,
    "type": "view",
  },
]
`;

exports[`remax render renders conditional fragment correctly 2`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "foo",
        "type": Symbol(text),
      },
    ],
    "id": 1,
    "props": Object {},
    "text": undefined,
    "type": "view",
  },
]
`;

exports[`remax render renders style 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "text": "hello",
        "type": Symbol(text),
      },
    ],
    "id": 1,
    "props": Object {
      "style": "width:100px;height:100px;",
    },
    "text": undefined,
    "type": "view",
  },
]
`;

exports[`remax render umount component 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "text": "foo",
            "type": Symbol(text),
          },
        ],
        "id": 1,
        "props": Object {},
        "text": undefined,
        "type": "view",
      },
    ],
    "id": 2,
    "props": Object {},
    "text": undefined,
    "type": "view",
  },
]
`;

exports[`remax render umount component 2`] = `
Array [
  Object {
    "children": Array [],
    "id": 2,
    "props": Object {},
    "text": undefined,
    "type": "view",
  },
]
`;
